File: linkedlist/leetcode_Copy_Linked_List_With_Random_Pointer.java
/*leetcode_Copy_Linked_List_With_Random_Pointer
leetcode/Microsoft

Given a linkedlist which has next() and random().
next() points to next node and random() points to a random node in this
linkedlist.

Copy this linkedlist in O(n)


A linked list is given such that each node contains an additional random
pointer which could point to any node in the list or null.

Return a deep copy of the list.
*/


class MS_Copy_Linked_List_With_Random_Pointer {

  public static void main(String[] args) {

  }

  public static ListNode copy(ListNode head) {
	if (head == null) {
	  return null;
	}

	RandomListNode cursor = head;
	// Given a -> b -> c -> null
	// Create a -> a -> b -> b -> c-> c -> null
	do {
	  RandomListNode node = cursor.next;
	  cursor.next = new RandomListNode(cursor.label);
	  cursor.next.next = node;
	  cursor = node;
	} while (cursor != null);

	// Point the random ref of copied node to copied.
	cursor = head;
	do {
	  cursor.next.random = (cursor.random == null) ? null : cursor.random.next;
	  cursor = cursor.next.next;
	} while (cursor != null);

	// Strip into two linkedlists.
	cursor = head;
	RandomListNode ret = cursor.next;

	for (RandomListNode n = ret; ; ) {
	  cursor.next = n.next;
	  cursor = cursor.next;

	  if (cursor == null) { break; }
	  n.next = cursor.next;
	  n = n.next;
	}

	return ret;
  }

  static class RandomListNode {
	RandomListNode next;
	RandomListNode random;
	int label;

	RandomListNode(int label) {
	  this.label = label;
	  this.next = null;
	  this.random = null;
	}
  }
}

File: linkedlist/leetcode_Copy_Linked_List_With_Random_Pointer.py
"""Copy linked list with random pointer
leetcode

A linked list is given such that each node contains an additional random pointer
which could point to any node in the list or null.

Return a deep copy of the list.
"""


# Definition for singly-linked list with a random pointer.
class RandomListNode(object):
    def __init__(self, x):
        self.label = x
        self.next = None
        self.random = None


class Solution(object):
    def copyRandomList(self, head):
        """
        :type head: RandomListNode
        :rtype: RandomListNode
        """
        if not head:
            return head
        cursor = head
        # Create a copy after each node in given list.
        while cursor is not None:
            nex = cursor.next
            node = RandomListNode(cursor.label)
            cursor.next = node
            node.next = nex
            cursor = nex

        # Copy the pointers.
        cursor = head
        while cursor is not None:
            if cursor.random:
                cursor.next.random = cursor.random.next
            cursor = cursor.next.next

        cursor = head
        copy_head = head.next
        copy_cursor = copy_head

        # Seperate original list and copied ones.
        while copy_cursor is not None:
            cursor.next = copy_cursor.next
            if cursor.next:
                copy_cursor.next = cursor.next.next

            cursor = cursor.next
            copy_cursor = copy_cursor.next
        return copy_head


def main():
    sol = Solution()


if __name__ == '__main__':
    main()
